sharktank/tests/models/llama/sharded_llama_test.py

```diff
@@ -18,6 +27,92 @@
+    hal_driver = iree.runtime.get_driver("local-task")
+    available_devices = hal_driver.query_available_devices()
+    # Use the same actual device for all devices.
+    return [hal_driver.create_device(available_devices[0]) for _ in range(device_count)]
+
+
+def load_iree_module(
+    module_path: str,
+    parameters_path: str,
+    devices: List[iree.runtime.HalDevice],
+) -> Tuple[iree.runtime.VmModule, iree.runtime.VmContext, iree.runtime.VmInstance]:
+    params_path = Path(parameters_path)
+    # TODO: make IREE able to load the parameters from the top parameter file
+    # without having to specify the parameter file for each shard separately.
+    parameter_index = iree.runtime.ParameterIndex()
+    for i in range(len(devices)):
+        parameter_index.load(
+            file_path=str(
+                Path(params_path).with_suffix(f".rank{i}{params_path.suffix}")
+            )
+        )
+    parameter_provider = parameter_index.create_provider(scope="model")
+    vm_instance = iree.runtime.VmInstance()
+    parameters_module = iree.runtime.create_io_parameters_module(
+        vm_instance, parameter_provider
+    )
+    vm_module = iree.runtime.VmModule.mmap(vm_instance, str(module_path))
+    hal_module = iree.runtime.create_hal_module(instance=vm_instance, devices=devices)
+    vm_context = iree.runtime.VmContext(
+        instance=vm_instance, modules=(hal_module, parameters_module, vm_module)
+    )
+    return vm_module, vm_context, vm_instance
+
+
+def run_iree_module_function(
+    module: iree.runtime.VmModule,
+    vm_context: iree.runtime.VmContext,
+    function_name: str,
+    args: List[iree.runtime.DeviceArray],
+) -> List[iree.runtime.DeviceArray]:
+    vm_function = module.lookup_function(function_name)
+    invoker = iree.runtime.FunctionInvoker(
+        vm_context=vm_context,
+        # TODO: rework iree.runtime.FunctionInvoker interface for multiple devices.
+        # This works, but does not look right.
+        device=iree.runtime.get_device("local-task", cache=False),
+        vm_function=vm_function,
+    )
+    res = invoker(*args)
+    if isinstance(res, iree.runtime.DeviceArray):
+        res = (res,)
+    return res
+
+
+def prepare_iree_module_function_args(
+    args: List[Union[AnyTensor, List[AnyTensor]]], devices: List[iree.runtime.HalDevice]
+) -> List[iree.runtime.DeviceArray]:
+    res = []
+    for arg in args:
+        if isinstance(arg, ShardedTensor):
+            assert len(devices) == len(arg.shards)
+            res.extend(
+                [
+                    prepare_iree_module_function_args([shard], [device])[0]
+                    for shard, device in zip(arg.shards, devices)
+                ]
+            )
+        elif isinstance(arg, (InferenceTensor, torch.Tensor)):
+            res.append(
+                iree.runtime.asdevicearray(
+                    devices[0], unbox_tensor(arg).to("cpu").numpy()
```

Comment:
Why do we need to unbox the tensor args here? Will become an issue for quantized tensors

Reply:
It can be `DeafultPrimitiveTensor`. I make the type check more specific.